csm_api_client/service/cfs.py

```diff
@@ -22,158 +22,492 @@
 # OTHER DEALINGS IN THE SOFTWARE.
 #
 """
-Client for querying the Configuration Framework Service (CFS) API
+Basic client library for CFS.
 """
-from datetime import datetime
+import os.path
+from datetime import datetime, timedelta
 import fnmatch
 from itertools import chain
+import json
+from enum import Enum
 import logging
 import re
+from typing import (
+    Dict,
+    Iterable,
+    List,
+    Optional,
+)
+from urllib.parse import urlparse, urlunparse
 import uuid
 
-from kubernetes.client import ApiException
+from cray_product_catalog.query import ProductCatalog, ProductCatalogError
 
-from sat.apiclient.gateway import APIError, APIGatewayClient
-from sat.util import get_val_by_path
+# ApiException is not found in the kubernetes.client library
```

Comment:
Ah, that was needed before I set `tool.mypy.ignore_missing_imports = true` in `pyproject.toml`. When that setting is `false`, `mypy` reports the following errors: 
```
csm_api_client/k8s.py:29: error: Skipping analyzing "kubernetes.client": module is installed, but missing library stubs or py.typed marker
csm_api_client/k8s.py:30: error: Skipping analyzing "kubernetes.config": module is installed, but missing library stubs or py.typed marker
csm_api_client/k8s.py:31: error: Skipping analyzing "kubernetes.config.config_exception": module is installed, but missing library stubs or py.typed marker
csm_api_client/service/vcs.py:45: error: Skipping analyzing "kubernetes.config": module is installed, but missing library stubs or py.typed marker
csm_api_client/service/vcs.py:46: error: Skipping analyzing "kubernetes.client": module is installed, but missing library stubs or py.typed marker
csm_api_client/session.py:40: error: Skipping analyzing "requests_oauthlib": module is installed, but missing library stubs or py.typed marker
csm_api_client/service/cfs.py:44: error: Skipping analyzing "cray_product_catalog.query": module is installed, but missing library stubs or py.typed marker
csm_api_client/service/cfs.py:44: note: See https://mypy.readthedocs.io/en/stable/running_mypy.html#missing-imports
csm_api_client/service/cfs.py:47: error: Skipping analyzing "kubernetes.client": module is installed, but missing library stubs or py.typed marker
Found 8 errors in 4 files (checked 10 source files)
```
These errors are caused by the listed libraries not shipping including type hints, and not having corresponding type stub packages (see [PEP 561](https://peps.python.org/pep-0561/).) It is possible to create minimal type stubs using `stubgen` for these libraries which only include annotations for the classes and functions we use, but this seemed outside the scope of this ticket. (See https://mypy.readthedocs.io/en/stable/stubgen.html for more details.)

If `mypy` considers an import "missing" (i.e. it cannot find type information for it) and the setting to ignore missing imports is enabled, it assigns it the dynamic type `Any` to the missing imports. Since the `Any` type can have inhabitants of any type, it isn't useful for the type checker since it can't determine if a given value is the wrong type. (e.g. if we have some function `foo(k8s: CoreV1Api)`, since `CoreV1Api` is a missing import, this function signature is equivalent to `foo(k8s: Any)`, for which any call would type-check, including `foo(1)`, `foo("string")`, etc.) However, for the sake of getting the change done, I opted to ignore missing imports for now, at the expense of some possible type unsoundness.

There is more information about how `mypy` handles missing imports here, which might help clarify all this information: https://mypy.readthedocs.io/en/stable/running_mypy.html#missing-imports